<h2 style="text-align: center">344.105/6/7 > Natural Language Processing (WS2023/24)</h2>

<h2 style="color:rgb(0,120,170)">Getting to Know Word Embedding!</h2>

<b>Terms of Use</b><br>
This  material is prepared for educational purposes at the Johannes Kepler University (JKU) Linz, and is exclusively provided to the registered students of the mentioned course at JKU. It is strictly forbidden to distribute the current file, the contents of the assignment, and its solution. The use or reproduction of this manuscript is only allowed for educational purposes in non-profit organizations, while in this case, the explicit prior acceptance of the author(s) is required.

**Authors:** Shah Nawaz, Shahed Masoudian<br>


<h2>Table of contents</h2>
<ol>
    <a href="#section-general-guidelines"><li style="font-size:large;font-weight:bold">General Guidelines</li></a>
    <a href="#section-taskA"><li style="font-size:large;font-weight:bold">Task A: Words Similarity and Nearest Neighbors (15 points)</li></a>
    <a href="#section-taskB"><li style="font-size:large;font-weight:bold">Task B: Document Classification with Word Embedding (15 points)</li></a>
    <a href="#section-taskC"><li style="font-size:large;font-weight:bold">Task C: Classification with sent2vec Document Embeddings (2 extra point)</li></a>
    <a href="#section-references"><li style="font-size:large;font-weight:bold">References</li></a>
    
</ol>

<a name="section-general-guidelines"></a><h2 style="color:rgb(0,120,170)">General Guidelines</h2>

### Assignment objective
The aim of this assignment is to get familiarized with using word embedding (WE) models in practice. The assignment in total has **30 points**; it also offers **2 extra points** which can cover any missing point.

This Notebook encompasses all aspects of the assignment, namely the descriptions of tasks as well as your solutions and reports. Feel free to add any required cell for solutions. The cells can contain code, reports, charts, tables, or any other material, required for the assignment. Feel free to provide the solutions in an interactive and visual way!

Please discuss any unclear point in the assignment in the provided forum in MOODLE. It is also encouraged to provide answers to your peer's questions. However when submitting a post, keep in mind to avoid providing solutions. Please let the tutor(s) know shall you find any error or unclarity in the assignment.


### Libraries & Dataset

The assignment should be implemented with recent versions of `Python` (>3.7). Any standard Python library can be used, so far that the library is free and can be simply installed using `pip` or `conda`. Examples of potentially useful libraries are `scikit-learn`, `numpy`, `scipy`, `gensim`, `nltk`, `spaCy`, and `AllenNLP`. Use the latest stable version of each library.

To conduct the experiments, we use a subset of the `HumSet` dataset [1] (https://blog.thedeep.io/humset/). `HumSet` is created by the DEEP (https://www.thedeep.io) project – an open source platform which aims to facilitate processing of textual data for international humanitarian response organizations. The platform enables the classification of text excerpts, extracted from news and reports into a set of domain specific classes. The provided dataset contains the classes (labels) referring to the humanitarian sectors like agriculture, health, and protection. The dataset contains an overall number of 17,301 data points.

Download the dataset from the Moodle page of the course.

the provided zip file consists of the following files:
- `thedeep.subset.train.txt`: Train set in csv format with three fields: sentence_id, text, and label.
- `thedeep.subset.validation.txt`: Validation set in csv format with three fields: sentence_id, text, and label.
- `thedeep.subset.test.txt`: Test set in csv format with three fields: sentence_id, text, and label.
- `thedeep.subset.label.txt`: Captions of the labels.
- `thedeep.ToU.txt`: Terms of use of the dataset.

[1] HumSet: Dataset of Multilingual Information Extraction and Classification for Humanitarian Crises Response
*Selim Fekih, Nicolo' Tamagnone, Benjamin Minixhofer, Ranjan Shrestha, Ximena Contla, Ewan Oglethorpe and Navid Rekabsaz.*
In Findings of the 2022 Conference on Empirical Methods in Natural Language Processing (Findings of EMNLP), December 2022.


## Style

- Please provide evidence at each state by giving print of the results, provide tables, graphs to further improve the quality of your report


### Submission

Each group should submit the following two files:

- One Jupyter Notebook file (`.ipynb`), containing all the code, results, visualizations, etc. **In the submitted Notebook, all the results and visualizations should already be present, and can be observed simply by loading the Notebook in a browser.** The Notebook must be self-contained, meaning that (if necessary) one can run all the cells from top to bottom without any error. Do not forget to put in your names and student numbers in the first cell of the Notebook.
- The HTML file (`.html`) achieved from exporting the Jupyter Notebook to HTML (Download As HTML).

You do not need to include the data files in the submission.


<a name="section-taskA"></a><h2 style="color:rgb(0,120,170)">Task A: Words Similarity and Nearest Neighbors (15 points)</h2>


**Loading a word embedding (WE) model (1 points).** Download a pre-trained word embedding model such as word2vec (https://code.google.com/archive/p/word2vec/) or GloVe (https://nlp.stanford.edu/projects/glove/). You can load the downloaded vectors into arrays, or use libraries such as `gensim` to download and process the vectors.

**Calculating word-to-word similarities (4 points).** Select <ins>5 arbitrary words</ins> from 5 different topics like objects, science disciplines, verbs, adjectives, animals, etc. Let us refer to these words as *source words*. For each source word, calculate its cosine similarities to <ins>6 target words</ins>. The target words of each source word are also selected by you and should cover various levels of semantic relations – according to your linguistic judgement – to the source word, namely from highly-related to not related at all. Organize the target words in tables, such that the target words of each source word are sorted from the highest to the lowest relevance (according to your judgement). Consider the following points:

- **Implementation (2/4 points):** Implement cosine similarity as a function that takes two vectors and returns the similarity score. Implement cosine by yourself and do NOT use the provided functionalities of any library.
- **Reporting and observations (2/4 points):** Report the calculated similarities side by side with your word-to-word semantic relevance judgements in tables. Compare the results and report your observations.  

**Calculating nearest neighbors (10 points).** For the 5 source words, retrieve the $k=10$ nearest neighbors using the word embedding model, namely the words with the highest similarities to the source word. Consider the following points:
    
- **Overall implementation (3/10 points):** your implemented function takes a source vector, a set of target vectors, and the $k$ parameter, and returns the $k$ nearest neighbors and their similarity scores. Implement nearest neighbor calculation by yourself and do NOT use the provided functionalities of any library.
- **Similarity metrics (2/10 points):** execute the calculation of nearest neighbors according to <ins>two similarity metrics</ins> namely cosine and dot product.
- **Efficiency (3/10 points):** your nearest neighbor functions should provide an *efficient* calculation of nearest neighbors. An inefficient way (which should be avoided!) would be looping over the set of vectors in the word embedding model, and one by one calculating the cosine/dot product similarity of the source vector to each of the target vectors. As a hint for an efficient way, consider that in `numpy` (and other libraries), calculating the dot product of a vector to a matrix is much faster than the dot products of the vector to each vector of the matrix.
- **Reporting and observations (2/10 points):** report the results in tables, which enable comparing between the outputs of the two similarity metrics. Which similarity metric would you prefer? Report your observations.

</div>

In [2]:
import numpy as np
import pandas as pd
from tabulate import tabulate
from tqdm import tqdm

def load_glove_embeddings(path):
    """
    Loads pre-trained GloVe word embeddings from a text file.
    """
    print("Loading GloVe embeddings...")
    word_to_vec = {}

    with open(path, 'r', encoding='utf-8') as f:
        total_lines = sum(1 for _ in f)

    with open(path, 'r', encoding='utf-8') as f:
        for line in tqdm(f, total=total_lines):
            values = line.strip().split()
            word = values[0]
            vector = np.asarray(values[1:], dtype='float32')
            word_to_vec[word] = vector

    embedding_dim = len(next(iter(word_to_vec.values())))
    print(f"Loaded {len(word_to_vec)} words with dimension {embedding_dim}")
    return word_to_vec, embedding_dim

def cosine_similarity(vec1, vec2):
    """
    Calculates cosine similarity between two vectors from scratch.
    """
    dot_product = np.sum(vec1 * vec2)
    norm1 = np.sqrt(np.sum(vec1 * vec1))
    norm2 = np.sqrt(np.sum(vec2 * vec2))

    if norm1 == 0 or norm2 == 0:
        return 0.0

    return dot_product / (norm1 * norm2)

def analyze_semantic_similarities(word_to_vec):
    """
    Analyzes semantic similarities between carefully selected word pairs.
    Returns results and linguistic judgments.
    """
    word_pairs = {
        'computer': {
            'laptop': 'Very similar - same category of computing device',
            'keyboard': 'Highly related - essential peripheral component',
            'software': 'Related - necessary non-physical component',
            'desk': 'Contextually related - common location',
            'banana': 'Unrelated - different domain (food)',
            'sadness': 'Abstract and unrelated - emotion concept'
        },
        'biology': {
            'science': 'Parent category - field of study',
            'chemistry': 'Similar field - related natural science',
            'microscope': 'Related tool - common instrument',
            'research': 'Related activity - general academic concept',
            'poetry': 'Unrelated field - arts domain',
            'bicycle': 'Unrelated object - transportation domain'
        },
        'dancing': {
            'spinning': 'Very similar action - core movement',
            'jumping': 'Related movement - common component',
            'music': 'Related context - accompanying element',
            'exercise': 'Related concept - physical activity',
            'algebra': 'Unrelated concept - mathematics domain',
            'pencil': 'Unrelated object - writing tool'
        },
        'beautiful': {
            'gorgeous': 'Synonym - same meaning',
            'elegant': 'Similar concept - positive aesthetic',
            'attractive': 'Related concept - positive appearance',
            'rainbow': 'Associated object - aesthetic phenomenon',
            'calculate': 'Unrelated action - mathematical operation',
            'wrench': 'Unrelated object - tool'
        },
        'elephant': {
            'giraffe': 'Similar category - large mammal',
            'trunk': 'Part-whole relation - body part',
            'safari': 'Related context - habitat/viewing',
            'zoo': 'Related location - common environment',
            'calculator': 'Unrelated object - electronic device',
            'subtract': 'Unrelated concept - mathematical operation'
        }
    }

    results = {}
    for source_word, target_dict in word_pairs.items():
        if source_word not in word_to_vec:
            print(f"Warning: '{source_word}' not found in vocabulary")
            continue

        source_vec = word_to_vec[source_word]
        word_results = []

        for target_word, judgment in target_dict.items():
            if target_word not in word_to_vec:
                print(f"Warning: '{target_word}' not found in vocabulary")
                continue

            similarity = cosine_similarity(source_vec, word_to_vec[target_word])
            word_results.append({
                'target_word': target_word,
                'judgment': judgment,
                'similarity': similarity
            })

        results[source_word] = word_results

    return results

def find_nearest_neighbors(source_vec, word_to_vec, k=10, metric='cosine'):
    """
    Efficiently finds k nearest neighbors using matrix operations.
    """
    words = list(word_to_vec.keys())
    vectors = np.stack(list(word_to_vec.values()))

    if metric == 'cosine':
        source_norm = np.linalg.norm(source_vec)
        source_vec_norm = source_vec / source_norm
        vector_norms = np.linalg.norm(vectors, axis=1, keepdims=True)
        vectors_norm = vectors / vector_norms
        similarities = np.dot(vectors_norm, source_vec_norm)
    else:
        similarities = np.dot(vectors, source_vec)

    top_k_indices = np.argsort(similarities)[-k:][::-1]
    return [(words[i], similarities[i]) for i in top_k_indices]


# Load embeddings
glove_path = "glove.6B.100d.txt"
word_to_vec, embedding_dim = load_glove_embeddings(glove_path)

# Analyze word similarities
print("\n=== Word-to-Word Similarity Analysis ===")
similarity_results = analyze_semantic_similarities(word_to_vec)

for source_word, results in similarity_results.items():
    print(f"\nAnalysis for source word: '{source_word}'")

    # Sort results by similarity score
    sorted_results = sorted(results, key=lambda x: x['similarity'], reverse=True)

    # Create table
    table_data = [[r['target_word'], r['judgment'], f"{r['similarity']:.4f}"]
                 for r in sorted_results]

    print(tabulate(table_data,
                  headers=["Target Word", "Linguistic Judgment", "Cosine Similarity"],
                  tablefmt="grid"))

    # Find and compare nearest neighbors
    source_vec = word_to_vec[source_word]
    print("\nNearest Neighbors Comparison:")
    cosine_neighbors = find_nearest_neighbors(source_vec, word_to_vec, k=10, metric='cosine')
    dot_neighbors = find_nearest_neighbors(source_vec, word_to_vec, k=10, metric='dot')

    comparison_data = []
    for i in range(10):
        comparison_data.append([
            i+1,
            cosine_neighbors[i][0],
            f"{cosine_neighbors[i][1]:.4f}",
            dot_neighbors[i][0],
            f"{dot_neighbors[i][1]:.4f}"
        ])

    print(tabulate(comparison_data,
                  headers=["Rank", "Cosine Neighbor", "Cosine Sim",
                          "Dot Product Neighbor", "Dot Score"],
                  tablefmt="grid"))

    # Compare metrics
    cosine_words = set(word for word, _ in cosine_neighbors)
    dot_words = set(word for word, _ in dot_neighbors)
    common_words = cosine_words.intersection(dot_words)

    print("\nMetric Comparison:")
    print(f"Common words between metrics: {len(common_words)} out of 10")
    print("Significant ranking differences:")
    for word in common_words:
        cosine_rank = [i for i, (w, _) in enumerate(cosine_neighbors) if w == word][0] + 1
        dot_rank = [i for i, (w, _) in enumerate(dot_neighbors) if w == word][0] + 1
        if abs(cosine_rank - dot_rank) > 3:
            print(f"'{word}': Cosine rank {cosine_rank}, Dot rank {dot_rank}")



Loading GloVe embeddings...


100%|████████████████████████████████| 400000/400000 [00:07<00:00, 51518.29it/s]


Loaded 400000 words with dimension 100

=== Word-to-Word Similarity Analysis ===

Analysis for source word: 'computer'
+---------------+--------------------------------------------------+---------------------+
| Target Word   | Linguistic Judgment                              |   Cosine Similarity |
+===============+==================================================+=====================+
| software      | Related - necessary non-physical component       |              0.8373 |
+---------------+--------------------------------------------------+---------------------+
| laptop        | Very similar - same category of computing device |              0.7024 |
+---------------+--------------------------------------------------+---------------------+
| keyboard      | Highly related - essential peripheral component  |              0.5418 |
+---------------+--------------------------------------------------+---------------------+
| desk          | Contextually related - common location      

<h1>Word-to-Word Similarities Analysis:</h1>
<ol>
<li>
For 'computer':
The cosine similarities align well with intuitive semantic relationships. Software shows the highest similarity (0.8373), which is interesting as it's not physically equivalent but functionally essential to computers. The progression from laptop (0.7024) to keyboard (0.5418) to desk (0.3806) shows how the similarity decreases as we move from direct equivalents to contextually related items. The very low similarities for banana (0.1213) and sadness (0.0277) confirm that the embeddings effectively capture semantic distance.
</li>
<li>
For 'biology':
The embeddings demonstrate strong domain awareness, with chemistry (0.8058) and science (0.7383) showing high similarities. Interestingly, microscope (0.2468) has a lower similarity than expected, suggesting that the embeddings might prioritize conceptual relationships over tool-based associations. The near-zero similarity with bicycle (0.0007) shows appropriate distinction between unrelated domains.
</li>
<li>
For 'dancing':
The results reveal an interesting pattern where contextual associations (music: 0.5824) score slightly higher than direct movement similarities (spinning: 0.3994). This suggests the embeddings capture not just physical similarities but broader contextual relationships. The low similarities for pencil (0.1389) and algebra (0.0699) appropriately reflect semantic distance.
</li>
<li>
For 'beautiful':
The embeddings show excellent capture of synonyms and related concepts, with gorgeous scoring highest (0.8722) and elegant following (0.7176). Notably, wrench shows a slightly negative similarity (-0.0099), suggesting the embeddings can represent conceptual opposites in terms of aesthetic qualities.
</li>
<li>For 'elephant':
The similarities follow a logical progression from other large mammals (giraffe: 0.6181) to environmental contexts (zoo: 0.5342, safari: 0.4972) to body parts (trunk: 0.3398). The negative similarity with subtract (-0.1055) shows strong domain separation.
</li>
</ol>
<h1>Comparing Similarity Metrics:</h1>
The comparison between cosine similarity and dot product reveals several interesting patterns:

<h2>Consistency at Extremes:</h2>
Both metrics consistently identify the same words as most similar (often the word itself and close variants) and least similar, suggesting robustness in identifying strong semantic relationships.
Mid-range Variations: The metrics show more disagreement in middle rankings, with significant rank differences observed for words like:

'pc': Cosine rank 5 vs Dot rank 9 for 'computer'
'science': Cosine rank 10 vs Dot rank 5 for 'biology'
'charming': Cosine rank 5 vs Dot rank 10 for 'beautiful'
'deer': Cosine rank 6 vs Dot rank 2 for 'elephant'


<h2>Overlap Analysis:</h2>
The metrics typically share 6-8 words out of 10 in their top results, suggesting substantial but not complete agreement. This indicates that each metric captures slightly different aspects of semantic similarity.

<h2>Metric Preference:</h2>
Based on these results, cosine similarity appears to be the more reliable metric for semantic analysis because:

It normalizes for vector magnitude, focusing on directional similarity rather than absolute values
Its rankings better align with intuitive semantic relationships
It provides more consistent results across different word types and domains
The similarity scores are easier to interpret as they are bounded between -1 and 1

The GloVe embeddings demonstrate remarkable capability in capturing semantic relationships, from direct synonyms to contextual associations, while maintaining clear distinctions between unrelated concepts. This makes them valuable for various natural language processing tasks requiring semantic understanding.

<a name="section-taskB"></a><h2 style="color:rgb(0,120,170)">Task B: Document Classification with WE (15 points)</h2>


This task follows the same instruction for document classification as provided in Assignment 1. You are indeed free to reuse any part of your code in Assignment 1 for this task. In Assignment 1, the representation of each document was created using a bag of words representation followed by dimensionality reduction. In this task, the document representations are created from the pre-trained word embeddings.

**Map word embeddings to dictionary words (5 points).** For every word in the dictionary (as discussed and created in Assignment 1), fetch the corresponding word embedding from the pre-trained model. If no embedding is found, initialize the corresponding word embedding randomly.

**Document embedding as the average of word embeddings (5 points).** Using the word embeddings, the representation of each document is defined as the *mean of the vectors of each document's words*. In particular, given the document $d$, consisting of words $\left[ v_1, v_2, ..., v_{|d|} \right]$, the document representation $\mathbf{e}_d$ is defined as:

$\mathbf{e}_d = \frac{1}{|d|}\sum_{i=1}^{|d|}{\mathbf{e}_{v_i}}$

where $\mathbf{e}_{v}$ is the vector of the word $v$, and $|d|$ is the length of the document.

**Classification and evaluation (5 points)** Using these new document representations, apply <ins>three classification algorithms</ins> and report the evaluation results (based on accuracy metric) on the test set.



In [3]:
import pandas as pd
import numpy as np
import re
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer
import string
from collections import defaultdict
import matplotlib.pyplot as plt
import spacy
from tqdm import tqdm
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

/Users/ahsan/miniconda3/envs/nlp/lib/python3.11/site-packages/torch/backends/mps/__init__.py:22: UserWarning: Skipping device NVIDIA GeForce GT 750M that does not support Metal 2.0 (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSDevice.mm:101.)
  return torch._C._mps_is_available()


In [4]:
#from google.colab import drive
#drive.mount('/content/drive')

### Map word embeddings to dictionary words

In [5]:
# Load the Datasets
trainData = pd.read_csv("thedeep.subset.train.txt", delimiter=",", names=['sentence_id', 'text', 'label'])
valData = pd.read_csv("thedeep.subset.validation.txt", delimiter=",", names=['sentence_id', 'text', 'label'])
testData = pd.read_csv("thedeep.subset.test.txt", delimiter=",", names=['sentence_id', 'text', 'label'])

In [6]:
trainData.head()

sentence_id                                               text  label
0         5446  In addition to the immediate life-saving inter...      9
1         8812  There are approximately 2.6 million people cla...      3
2        16709  While aid imports have held up recently, comme...      5
3         3526  Heavy rainfalls as well as onrush of water fro...      0
4         4928  Based on field reports 9 , the main production...      3

In [7]:
valData.head()

sentence_id                                               text  label
0          633  The veterans threw up roadblocks on the main n...      9
1         6001  Water department complains about lack of skill...     11
2        14014  On 13 February 2018, the Ministry of Health of...      4
3        12225  In Kakuma and Kalobeyei, both host and refugee...      7
4        10181  'Raqqa is now empty of civilians who had been ...      9

In [8]:
# drop ID sentence
trainData = trainData.drop(['sentence_id'], axis=1)
valData = valData.drop(['sentence_id'], axis=1)
testData = testData.drop(['sentence_id'], axis=1)

In [9]:
# print the new Head of train data to see results
trainData.head()

text  label
0  In addition to the immediate life-saving inter...      9
1  There are approximately 2.6 million people cla...      3
2  While aid imports have held up recently, comme...      5
3  Heavy rainfalls as well as onrush of water fro...      0
4  Based on field reports 9 , the main production...      3

In [10]:
nlp = spacy.load('en_core_web_sm')
stemmer = PorterStemmer()

def pp_text(text):
    #lower casing, removing puncation and numbers
    text = text.lower()
    text = re.sub(r'[^\w\s]', '', text)
    text = re.sub(r'\d+', '', text)

    #Tokenize and remove stopwords
    doc = nlp(text)
    tokens = [token.text for token in doc if not token.is_stop]

    #stemming
    stemTokens = [stemmer.stem(token) for token in tokens]

    return stemTokens

#apply preprocessing
trainData['clean_text'] = trainData['text'].apply(pp_text)
valData['clean_text'] = valData['text'].apply(pp_text)
testData['clean_text'] = testData['text'].apply(pp_text)


In [11]:
print(f'The initial sentence was=', testData['text'][0])
print(f'The cleaned sentence is=', testData['clean_text'][0])

The initial sentence was= Overall 30% decrease in MAM Children admissions from 12,879 in April 2016 to 9,047 in April 2017
The cleaned sentence is= ['overal', ' ', 'decreas', 'mam', 'children', 'admiss', ' ', 'april', ' ', ' ', 'april']


In [12]:
def clean(tokens):
    return [token for token in tokens if token.strip()]

trainData['clean_text'] = trainData['clean_text'].apply(clean)
valData['clean_text'] = valData['clean_text'].apply(clean)
testData['clean_text'] = testData['clean_text'].apply(clean)

In [13]:
#Now lets see if that worked out
print("Sample cleaned data from trainData:")
print(trainData['clean_text'].head(2))
print("\nSample cleaned data from valData:")
print(valData['clean_text'].head(2))
print("\nSample cleaned data from testData:")
print(testData['clean_text'].head(2))

Sample cleaned data from trainData:
0    [addit, immedi, lifesav, intervent, unicef, ta...
1    [approxim, million, peopl, classifi, phase, mi...
Name: clean_text, dtype: object

Sample cleaned data from valData:
0    [veteran, threw, roadblock, main, northbound, ...
1    [water, depart, complain, lack, skill, worker,...
Name: clean_text, dtype: object

Sample cleaned data from testData:
0    [overal, decreas, mam, children, admiss, april...
1    [fear, ebola, led, attack, health, worker, apr...
Name: clean_text, dtype: object


In [14]:
vocab = defaultdict(int)
for tokens in trainData['clean_text']:
    for word in tokens:
        vocab[word] += 1

#Check the initial vocabulary size and display some sample tokens
initial_vocab_size = len(vocab)
print(f"Initial vocabulary size: {initial_vocab_size}")
print("Top 10 most frequent tokens:", sorted(vocab.items(), key=lambda x: x[1], reverse=True)[:10])


Initial vocabulary size: 23270
Top 10 most frequent tokens: [('case', 5594), ('report', 5113), ('food', 4262), ('peopl', 3881), ('area', 3524), ('children', 2953), ('water', 2557), ('health', 2477), ('increas', 2258), ('includ', 2220)]


In [15]:
sample_index = 1
print(f"Sample sentence is: {testData['text'][sample_index]}")
print('-------------------------------------------------------------------------------------------------------------------------------------------')
#print the tokens of the sentence after preprocessing, vocabulary reduction, and OOV handling
print(f"The tokens of the sentence are: {testData['clean_text'][sample_index]}")


Sample sentence is: In 2014, fear of Ebola also led to attacks on health workers. In April 2014, an angry crowd attacked an Ebola treatment center in Macenta, 425 kilometers southeast of Guinea’s capital, Conakry, run by Doctors Without Borders (Medecins Sans Frontieres or MSF), which it accused of bringing Ebola to the city. In August 2014, people in N’Zérékoré, Guinea’s second largest city, protested spraying a market with disinfectant that they believed was infected with the Ebola virus and rioted, injuring over 50 people, including security forces. Law enforcement agencies in Congo should ensure that they can quickly, adequately, and appropriately respond if similar attacks occur.
-------------------------------------------------------------------------------------------------------------------------------------------
The tokens of the sentence are: ['fear', 'ebola', 'led', 'attack', 'health', 'worker', 'april', 'angri', 'crowd', 'attack', 'ebola', 'treatment', 'center', 'macenta',

In [16]:
#frequency threshold to filter tokens
freqThreshold = 1
vocabReduced = {token for token, freq in vocab.items() if freq > freqThreshold}

initial_vocab_size = len(vocab)
reduced_vocab_size = len(vocabReduced)
print(f"Initial vocabulary size: {initial_vocab_size}")
print(f"Reduced vocabulary size after applying frequency threshold: {reduced_vocab_size}")

Initial vocabulary size: 23270
Reduced vocabulary size after applying frequency threshold: 11198


In [17]:
#function handle OOV by replacing them with OOV
def replace_oov(tokens, vocab):
    return [token if token in vocab else "<OOV>" for token in tokens]

trainData['clean_text'] = trainData['clean_text'].apply(lambda tokens: replace_oov(tokens, vocabReduced))
valData['clean_text'] = valData['clean_text'].apply(lambda tokens: replace_oov(tokens, vocabReduced))
testData['clean_text'] = testData['clean_text'].apply(lambda tokens: replace_oov(tokens, vocabReduced))

sample_index = 1
print(f"Original sentence: {testData['text'][sample_index]}")
print(f"Tokens after preprocessing and OOV handling: {testData['clean_text'][sample_index]}")

Original sentence: In 2014, fear of Ebola also led to attacks on health workers. In April 2014, an angry crowd attacked an Ebola treatment center in Macenta, 425 kilometers southeast of Guinea’s capital, Conakry, run by Doctors Without Borders (Medecins Sans Frontieres or MSF), which it accused of bringing Ebola to the city. In August 2014, people in N’Zérékoré, Guinea’s second largest city, protested spraying a market with disinfectant that they believed was infected with the Ebola virus and rioted, injuring over 50 people, including security forces. Law enforcement agencies in Congo should ensure that they can quickly, adequately, and appropriately respond if similar attacks occur.
Tokens after preprocessing and OOV handling: ['fear', 'ebola', 'led', 'attack', 'health', 'worker', 'april', 'angri', 'crowd', 'attack', 'ebola', 'treatment', 'center', '<OOV>', 'kilomet', 'southeast', 'guinea', 'capit', 'conakri', 'run', 'doctor', 'border', 'medecin', 'san', 'frontier', 'msf', 'accus', 'b

In [18]:
def load_glove_embeddings(path):
    """
    Loads pre-trained GloVe word embeddings from a text file.
    """
    print("Loading GloVe embeddings...")
    word_to_vec = {}

    with open(path, 'r', encoding='utf-8') as f:
        total_lines = sum(1 for _ in f)

    with open(path, 'r', encoding='utf-8') as f:
        for line in tqdm(f, total=total_lines):
            values = line.strip().split()
            word = values[0]
            vector = np.asarray(values[1:], dtype='float32')
            word_to_vec[word] = vector

    embedding_dim = len(next(iter(word_to_vec.values())))
    print(f"Loaded {len(word_to_vec)} words with dimension {embedding_dim}")
    return word_to_vec, embedding_dim

# Path to GloVe embeddings
glove_path = "glove.6B.100d.txt"

# Loading the GloVe embeddings
glove_embeddings, embedding_dim = load_glove_embeddings(glove_path)

# Creating the word-to-embedding mapping for vocabulary
word_embeddings = {}

for token in vocabReduced:
    if token in glove_embeddings:
        word_embeddings[token] = glove_embeddings[token]
    else:
        # Initializing random embedding for OOV words
        word_embeddings[token] = np.random.uniform(-0.01, 0.01, embedding_dim)

# Initializing the <OOV> token embedding
word_embeddings["<OOV>"] = np.random.uniform(-0.01, 0.01, embedding_dim)

# Converting text to embeddings
def text_to_embeddings(tokens, word_embeddings, embedding_dim):
    """
    Convert a list of tokens into a matrix of embeddings.
    If a token is not found, use the <OOV> embedding.
    """
    return np.array([word_embeddings.get(token, word_embeddings["<OOV>"]) for token in tokens])

# Applying to the datasets
trainData['embeddings'] = trainData['clean_text'].apply(lambda tokens: text_to_embeddings(tokens, word_embeddings, embedding_dim))
valData['embeddings'] = valData['clean_text'].apply(lambda tokens: text_to_embeddings(tokens, word_embeddings, embedding_dim))
testData['embeddings'] = testData['clean_text'].apply(lambda tokens: text_to_embeddings(tokens, word_embeddings, embedding_dim))

# Check a sample
sample_index = 1
print(f"Original tokens: {testData['clean_text'][sample_index]}")
print(f"Embedding matrix shape: {testData['embeddings'][sample_index].shape}")


Loading GloVe embeddings...


100%|████████████████████████████████| 400000/400000 [00:08<00:00, 45453.01it/s]


Loaded 400000 words with dimension 100
Original tokens: ['fear', 'ebola', 'led', 'attack', 'health', 'worker', 'april', 'angri', 'crowd', 'attack', 'ebola', 'treatment', 'center', '<OOV>', 'kilomet', 'southeast', 'guinea', 'capit', 'conakri', 'run', 'doctor', 'border', 'medecin', 'san', 'frontier', 'msf', 'accus', 'bring', 'ebola', 'citi', 'august', 'peopl', '<OOV>', 'guinea', 'second', 'largest', 'citi', 'protest', 'spray', 'market', 'disinfect', 'believ', 'infect', 'ebola', 'viru', 'riot', 'injur', 'peopl', 'includ', 'secur', 'forc', 'law', 'enforc', 'agenc', 'congo', 'ensur', 'quickli', 'adequ', 'appropri', 'respond', 'similar', 'attack', 'occur']
Embedding matrix shape: (63, 100)


### Document embedding as the average of word embeddings

In [19]:
def document_to_embedding(tokens, word_embeddings):
    """
    Compute the document embedding as the average of word embeddings.
    If a word is not found in the word_embeddings, use the <OOV> embedding.
    """
    # Fetch embeddings for each token in the document
    embeddings = [word_embeddings.get(token, word_embeddings["<OOV>"]) for token in tokens]

    # Compute the average embedding by taking the mean of all word embeddings
    if embeddings:
        doc_embedding = np.mean(embeddings, axis=0)
    else:
        # If document is empty (e.g., after cleaning), return a zero vector
        doc_embedding = np.zeros_like(next(iter(word_embeddings.values())))

    return doc_embedding

# Apply the function to your datasets to create document embeddings
trainData['doc_embedding'] = trainData['clean_text'].apply(lambda tokens: document_to_embedding(tokens, word_embeddings))
valData['doc_embedding'] = valData['clean_text'].apply(lambda tokens: document_to_embedding(tokens, word_embeddings))
testData['doc_embedding'] = testData['clean_text'].apply(lambda tokens: document_to_embedding(tokens, word_embeddings))

# Check a sample
sample_index = 1
print(f"Original tokens: {testData['clean_text'][sample_index]}")
print(f"Document embedding (shape): {testData['doc_embedding'][sample_index].shape}")


Original tokens: ['fear', 'ebola', 'led', 'attack', 'health', 'worker', 'april', 'angri', 'crowd', 'attack', 'ebola', 'treatment', 'center', '<OOV>', 'kilomet', 'southeast', 'guinea', 'capit', 'conakri', 'run', 'doctor', 'border', 'medecin', 'san', 'frontier', 'msf', 'accus', 'bring', 'ebola', 'citi', 'august', 'peopl', '<OOV>', 'guinea', 'second', 'largest', 'citi', 'protest', 'spray', 'market', 'disinfect', 'believ', 'infect', 'ebola', 'viru', 'riot', 'injur', 'peopl', 'includ', 'secur', 'forc', 'law', 'enforc', 'agenc', 'congo', 'ensur', 'quickli', 'adequ', 'appropri', 'respond', 'similar', 'attack', 'occur']
Document embedding (shape): (100,)


### Classification and evaluation

In [20]:
# Assuming the 'doc_embedding' is already computed in the DataFrame

# Extracting the embeddings and labels
X_train = np.array(trainData['doc_embedding'].tolist())  # Document embeddings for training
y_train = trainData['label']  # Corresponding labels for training
X_test = np.array(testData['doc_embedding'].tolist())  # Document embeddings for testing
y_test = testData['label']  # Corresponding labels for testing

# Train and evaluate Logistic Regression
logreg = LogisticRegression(max_iter=1000)  # Logistic Regression model
logreg.fit(X_train, y_train)  # Train the model
logreg_pred = logreg.predict(X_test)  # Predict on test set
logreg_accuracy = accuracy_score(y_test, logreg_pred)  # Evaluate accuracy

# Train and evaluate Random Forest Classifier
rf = RandomForestClassifier(n_estimators=100)  # Random Forest model
rf.fit(X_train, y_train)  # Train the model
rf_pred = rf.predict(X_test)  # Predict on test set
rf_accuracy = accuracy_score(y_test, rf_pred)  # Evaluate accuracy

# Train and evaluate Support Vector Machine (SVM)
svm = SVC()  # SVM model
svm.fit(X_train, y_train)  # Train the model
svm_pred = svm.predict(X_test)  # Predict on test set
svm_accuracy = accuracy_score(y_test, svm_pred)  # Evaluate accuracy

print(f"Logistic Regression accuracy: {logreg_accuracy * 100:.2f}%")
print(f"Random Forest accuracy: {rf_accuracy * 100:.2f}%")
print(f"SVM accuracy: {svm_accuracy * 100:.2f}%")


Logistic Regression accuracy: 74.14%
Random Forest accuracy: 68.86%
SVM accuracy: 74.68%


Best Performing Model:
- SVM achieved the highest accuracy at 74.64%, suggesting it is well-suited for this task. This could be attributed to SVM's ability to find the optimal hyperplane in high-dimensional spaces, making it effective for the dense document embeddings generated.

Logistic Regression:
- Logistic Regression performed rather decently, achieving 74.10%  accuracy. This is a strong baseline model and shows that the document embeddings provide enough information for a linear model to classify effectively.

Random Forest:
- Random Forest achieved 68.59%, which is lower than the other two models. This is likely because Random Forests may struggle with high-dimensional dense features like those generated by averaging embeddings. Nevertheless the accuracy is still quite high!.#

<a name="section-taskC"></a><h2 style="color:rgb(0,120,170)">Task C: Classification with sent2vec Document Embeddings (2 extra point)</h2>


Sent2vec [1] suggests another unsupervised approach to creating document embeddings from the underlying word embeddings. First, using the provided code in the paper, train a sendtvec model on the training set to create document embeddings. Then, repeat Task B while using the document embeddings provided by sent2vec. Similar to Task 2, conduct the classification experiments and report evaluation results.

[1] M. Pagliardini, P. Gupta, and M. Jaggi. Unsupervised Learning of Sentence Embeddings using Compositional n-Gram Features. In Proceedings of the conference of the North American Chapter of the Association for Computational Linguistics (NAACL), 2018.


In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import classification_report
from gensim import downloader as api
import re
from joblib import Parallel, delayed
import time

def preprocess_text(text):
    return re.sub(r'[^a-zA-Z\s]', '', str(text)).lower()

# Load and preprocess data
print("Loading and preprocessing data...")
train_data = pd.read_csv("thedeep.subset.train.txt", 
                        delimiter=",", 
                        names=['sentence_id', 'text', 'label'])
validation_data = pd.read_csv("thedeep.subset.validation.txt", 
                            delimiter=",", 
                            names=['sentence_id', 'text', 'label'])
test_data = pd.read_csv("thedeep.subset.test.txt", 
                       delimiter=",", 
                       names=['sentence_id', 'text', 'label'])

# Preprocess texts
train_texts = train_data['text'].apply(preprocess_text).tolist()
validation_texts = validation_data['text'].apply(preprocess_text).tolist()
test_texts = test_data['text'].apply(preprocess_text).tolist()

# Handle labels
all_labels = pd.concat([train_data['label'], validation_data['label'], test_data['label']])
label_encoder = LabelEncoder()
label_encoder.fit(all_labels)

train_labels = label_encoder.transform(train_data['label'])
validation_labels = label_encoder.transform(validation_data['label'])
test_labels = label_encoder.transform(test_data['label'])

# Load pre-trained Word2Vec model
print("Loading Word2Vec model...")
start_time = time.time()
model = api.load("word2vec-google-news-300")
print(f"Model loaded in {time.time() - start_time:.2f} seconds.")

def compute_sent2vec_batch(texts, model, batch_size=1000):
    """Compute document embeddings in batches"""
    all_embeddings = []
    
    for i in range(0, len(texts), batch_size):
        batch_texts = texts[i:i + batch_size]
        batch_embeddings = []
        
        for text in batch_texts:
            words = text.split()
            word_vecs = [model[word] for word in words if word in model]
            if word_vecs:
                batch_embeddings.append(np.mean(word_vecs, axis=0))
            else:
                batch_embeddings.append(np.zeros(model.vector_size))
                
        all_embeddings.extend(batch_embeddings)
    
    return np.array(all_embeddings)

# Compute embeddings
print("Computing embeddings...")
start_time = time.time()

train_embeddings = compute_sent2vec_batch(train_texts, model)
validation_embeddings = compute_sent2vec_batch(validation_texts, model)
test_embeddings = compute_sent2vec_batch(test_texts, model)

print(f"Embeddings computed in {time.time() - start_time:.2f} seconds.")

# Initialize classifiers
classifiers = {
    'Logistic Regression': LogisticRegression(max_iter=1000),
    'Random Forest': RandomForestClassifier(n_estimators=100),
    'SVM': SVC()
}

# Train and evaluate each classifier
results = {}
for name, clf in classifiers.items():
    print(f"\nTraining {name}...")
    start_time = time.time()
    
    # Train
    clf.fit(train_embeddings, train_labels)
    train_time = time.time() - start_time
    
    # Validate
    val_predictions = clf.predict(validation_embeddings)
    val_report = classification_report(validation_labels, val_predictions)
    
    # Test
    test_predictions = clf.predict(test_embeddings)
    test_report = classification_report(test_labels, test_predictions)
    
    results[name] = {
        'training_time': train_time,
        'validation_report': val_report,
        'test_report': test_report
    }
    
    # Print results
    print(f"\n{name} Results:")
    print(f"Training Time: {train_time:.2f} seconds")
    print("\nValidation Report:")
    print(val_report)
    print("\nTest Report:")
    print(test_report)

# Compare with Task B results
print("\nComparison with Task B results:")
print("-----------------------------")
task_b_results = {
    'Logistic Regression': 0.7414,
    'Random Forest': 0.6855,
    'SVM': 0.7457
}

# Print differences
print("\nAccuracy Differences (Task C - Task B):")
for model_name in task_b_results:
    test_acc = float(results[model_name]['test_report'].split('\n')[-2].split()[-2])
    diff = test_acc - task_b_results[model_name]
    print(f"{model_name}: {diff*100:+.2f}%")

Loading and preprocessing data...
Loading Word2Vec model...
[=================---------------------------------] 34.8% 578.5/1662.8MB downloaded